### PR TITLE
Id class entity views

### PR DESCRIPTION
**Pull request by jwgmeligmeyling**

### CODE REVIEW

File: entity-view/api/src/main/java/com/blazebit/persistence/view/metamodel/ViewType.java [hunk 1]
```diff
@@ -43,6 +43,13 @@
      */
     public MethodAttribute<? super X, ?> getIdAttribute();
```
Comment (jwgmeligmeyling) [inline entity-view/api/src/main/java/com/blazebit/persistence/view/metamodel/ViewType.java:44]: Is this method still used?

File: entity-view/api/src/main/java/com/blazebit/persistence/view/spi/EntityViewMapping.java [hunk 2]
```diff
@@ -194,6 +195,14 @@
      */
     public void setIdAttribute(EntityViewAttributeMapping idAttribute);
```
Comment (jwgmeligmeyling) [inline entity-view/api/src/main/java/com/blazebit/persistence/view/spi/EntityViewMapping.java:196]: Is this method still used?

File: entity-view/api/src/main/java/com/blazebit/persistence/view/spi/EntityViewMapping.java [hunk 3]
```diff
@@ -194,6 +195,14 @@
...
 
+    /**
+     * Set the id attribute mapping of this entity view mapping.
+     * Note that the attribute must be one of the attributes as given by {@link #getAttributes()} or <code>null</code>.
```
Comment (jwgmeligmeyling) [inline entity-view/api/src/main/java/com/blazebit/persistence/view/spi/EntityViewMapping.java:200]: Description should be changed to plural form

File: entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/EntityViewManagerImpl.java [hunk 4]
```diff
@@ -244,9 +244,56 @@ public ProxyFactory getProxyFactory() {
...
+//                field.setAccessible(Boolean.TRUE);
+//                try {
+//                    entityIdValues.put(field.getName(),field.get(entityId));
+//                } catch (IllegalAccessException e) {
+//                    throw new IllegalStateException("Could not access field: "+field.getName()+".");
+//                }
+//            }
+//            entityIdValues.remove("serialVersionUID");
+            for (SingularAttribute<?,?> idAttribute : idAttributeSet){
+                Method method = (Method) idAttribute.getJavaMember();
```
Comment (jwgmeligmeyling) [inline entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/EntityViewManagerImpl.java]: I think that the `Member` can also be of type `Field`, in which case you do not want to cast it to a `Method`, but rather cast it to a `Field`, and then do `field.get(entityId)` instead of `method.invoke(entityId)`. Actually, I am worried that theres even much more to it than that. See for example the code in `JpaMetamodelUtils.resolveFieldClass(basClass, attr)`, which looks up the type of the field of an attribute.

File: entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/EntityViewManagerImpl.java [hunk 5]
```diff
@@ -244,9 +244,56 @@ public ProxyFactory getProxyFactory() {
...
+            Iterator iterator = idAttributeSet.iterator();
+            if(!iterator.hasNext()){
+                throw new RuntimeException("The entity type" + entityType.getName() + "does not have an Id!");
+                }
+
+            //I believe this approach also works for BigDecimal and the like? I have added that possibility in the check above.
+            entityIdValues.put(((SingularAttribute) iterator.next()).getName(),entityId);
+
+            if (iterator.hasNext()){
+                throw new RuntimeException("Could not match the given entityId to the entity type specified in the view: the entity type"
```
Comment (jwgmeligmeyling) [inline entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/EntityViewManagerImpl.java]: This exception should never occur, because you implemented support for compound keys.

File: entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/EntityViewManagerImpl.java [hunk 6]
```diff
@@ -244,9 +244,56 @@ public ProxyFactory getProxyFactory() {
...
-                .where(idAttribute.getName()).eq(entityId);
+        Set<SingularAttribute<?, ?>> idAttributeSet = JpaMetamodelUtils.getIdAttributes(entityType);
+        CriteriaBuilder<?> cb = cbf.create(entityManager, managedViewType.getEntityClass());
+        Map<String,Object> entityIdValues = new HashMap<>();
+        //TODO Hibernate returns null when getIdType it is called on an entityType which uses IdClass.
+        if(entityType.getIdType()!=null && (entityType.getIdType().getPersistenceType().equals(Type.PersistenceType.BASIC)||
+                entityType.getIdType().getPersistenceType().equals(Type.PersistenceType.EMBEDDABLE))){
+            Iterator iterator = idAttributeSet.iterator();
+            if(!iterator.hasNext()){
+                throw new RuntimeException("The entity type" + entityType.getName() + "does not have an Id!");
```
Comment (jwgmeligmeyling) [inline entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/EntityViewManagerImpl.java]: Entities are required to have an id, so I don't think we have to account for this? If we do, we should also handle for it in the else branch.

File: entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/EntityViewManagerImpl.java [hunk 7]
```diff
@@ -244,9 +244,56 @@ public ProxyFactory getProxyFactory() {
...
         ViewTypeImpl<T> managedViewType = metamodel.view(entityViewSetting.getEntityViewClass());
         EntityType<?> entityType = (EntityType<?>) managedViewType.getJpaManagedType();
-        SingularAttribute<?, ?> idAttribute = JpaMetamodelUtils.getSingleIdAttribute(entityType);
-        CriteriaBuilder<?> cb = cbf.create(entityManager, managedViewType.getEntityClass())
-                .where(idAttribute.getName()).eq(entityId);
+        Set<SingularAttribute<?, ?>> idAttributeSet = JpaMetamodelUtils.getIdAttributes(entityType);
+        CriteriaBuilder<?> cb = cbf.create(entityManager, managedViewType.getEntityClass());
+        Map<String,Object> entityIdValues = new HashMap<>();
+        //TODO Hibernate returns null when getIdType it is called on an entityType which uses IdClass.
+        if(entityType.getIdType()!=null && (entityType.getIdType().getPersistenceType().equals(Type.PersistenceType.BASIC)||
```
Comment (jwgmeligmeyling) [inline entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/EntityViewManagerImpl.java]: Can't we detect this using `idAttributeSet.size() == 1`?

File: entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/EntityViewManagerImpl.java [hunk 8]
```diff
@@ -244,9 +244,56 @@ public ProxyFactory getProxyFactory() {
...
+            //I believe this approach also works for BigDecimal and the like? I have added that possibility in the check above.
+            entityIdValues.put(((SingularAttribute) iterator.next()).getName(),entityId);
+
+            if (iterator.hasNext()){
+                throw new RuntimeException("Could not match the given entityId to the entity type specified in the view: the entity type"
+                        + entityType.getName() + "has more than one Id field!");
+            }
+        } else {
+//            for (Field field : Arrays.asList(entityId.getClass().getDeclaredFields())){
+//                field.setAccessible(Boolean.TRUE);
```
Comment (jwgmeligmeyling) [inline entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/EntityViewManagerImpl.java]: I don't think we need this anymore, finding the field based on the provided member should probably work. If not, we can still lookup the field by attribute name, as these, according to the spec, should match as far as I remember.

File: entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/entity/FlusherBasedEntityLoader.java [hunk 9]
```diff
@@ -55,8 +62,10 @@ private String getQueryString() {
                 flushers[i].appendFetchJoinQueryFragment("e", sb);
             }
         }
-        sb.append(" WHERE e.").append(idAttributeName).append(" = :id");
-
+        for (String idAttributeName : idAttributeNames){
+            sb.append(" WHERE e.").append(idAttributeName).append(" = :"+idAttributeName.replace(".","_")+"_variable"+" AND");
```
Comment (jwgmeligmeyling) [inline entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/entity/FlusherBasedEntityLoader.java:66]: `idAttributeName.replace(".","_")+"_variable"` works because it's used below in the similar fashion. Please extract this to a method `getParameterName(String idAttributeName)`, so we have more safety the two remain using the same naming scheme.

File: entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/entity/FlusherBasedEntityLoader.java [hunk 10]
```diff
@@ -71,16 +80,69 @@ public Object toEntity(UpdateContext context, Object id) {
         return getReferenceOrLoad(context, id);
     }
 
+    public Map<String,Object> getIdNameValueMap(Class entityClass, Object id, EntityManager em){
+        @SuppressWarnings("unchecked")
```
Comment (jwgmeligmeyling) [inline entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/entity/FlusherBasedEntityLoader.java]: You don't need this do you?

File: entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/entity/FlusherBasedEntityLoader.java [hunk 11]
```diff
@@ -71,16 +80,69 @@ public Object toEntity(UpdateContext context, Object id) {
         return getReferenceOrLoad(context, id);
     }
 
+    public Map<String,Object> getIdNameValueMap(Class entityClass, Object id, EntityManager em){
```
Comment (jwgmeligmeyling) [inline entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/entity/FlusherBasedEntityLoader.java]: Try to pass the `Metamodel` rather than the `EntityManager`.

Comment (jwgmeligmeyling) [reply entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/entity/FlusherBasedEntityLoader.java]: I'd also use it as first parameter.

**Review by jwgmeligmeyling — COMMENTED** [review]
These are lots and lots of line changes, so I only got half way through yet. In general it looks good, I've placed a few remarks. I think these are necessairy to make the solution work for other mappings / ORMs as well. Nevertheless, its certainly going into the right direction.

One issue I do see is that "id attribute names" and "id attribute paths" get mixed up. What do I mean by this distinction: an entity may have multiple `@Id` attributes, in which case an `@IdClass` will exists, which will have *fields* that correspond to the underlying identifier types (or type it self, if it is a basic type).

When you're trying to create a `WHERE` clause for a particular identifiable, you need to match the identifier *path expressions* to the corresponding values from the `@IdClass` object. Currently, you incorrectly use the attribute name as path expression. This only works for basic types. For example, for associations, you'll get type exceptions. Your mapping would eventually lead to a where clause of the form `entityAlias.someAssociation = :someLongValue`. This will throw an error because the identifier type of the association (`Long`) does not match the association type. Instead, your path expression should be `entityAlias.someAssociation.id`, with the final where clause `entityAlias.someAssociation.id = :someLongValue`. This can even happen in nested cases, ie. its perfectly valid to construct a case that requires `entityAlias.someAssociation.someOtherAssociation.id`. See also my work on #571 for implementing this for binding associations as CTE values in https://github.com/Blazebit/blaze-persistence/commit/f8cc47b115faf6deb7f709a6dc07f2e362c54b25.

So in conclusion: `getIdNameValueMap` need some rethinking. What you actually want is a set of WHERE clauses, consisting of a path expression (i.e. a `String`) and a corresponding value. You could for example rename it  to `getIdentifierPathExpressionValueMapping` or something of that kind. Then it needs to be adjusted to handle the (possibly nested) associations.

**Comment by jwgmeligmeyling** [comment]
By the way, for now you can probably best put `getIdentifierPathExpressionValueMapping` as a static method in the `JpaMetamodelUtil`.
